src/repositories/updater.rs

```diff
@@ -0,0 +1,366 @@
+    /// `None` is returned.
+    fn fetch_repository(&self, name: &RepositoryName) -> Result<Option<Repository>>;
+
+    /// Used by update_all_crates.
+    ///
+    /// The returned struct will contain all the information needed for `RepositoriesUpdater` to
+    /// update repositories that are still present and delete the missing ones.
+    fn fetch_repositories(&self, ids: &[String]) -> Result<FetchRepositoriesResult>;
+}
+
+#[derive(Debug)]
+pub struct Repository {
+    pub id: String,
+    pub name_with_owner: String,
+    pub description: Option<String>,
+    pub last_activity_at: Option<DateTime<Utc>>,
+    pub stars: i64,
+    pub forks: i64,
+    pub issues: i64,
+}
+
+#[derive(Default, Debug)]
+pub struct FetchRepositoriesResult {
+    pub present: HashMap<String, Repository>,
+    pub missing: Vec<String>,
+}
+
+pub struct RepositoryStatsUpdater {
+    updaters: Vec<Box<dyn RepositoryForge + Send + Sync>>,
+    pool: Pool,
+}
+
+impl fmt::Debug for RepositoryStatsUpdater {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "RepositoryStatsUpdater {{ {:?} }}",
+            self.updaters.iter().map(|u| u.host()).collect::<Vec<_>>()
+        )
+    }
+}
+
+impl RepositoryStatsUpdater {
+    pub fn new(config: &Config, pool: Pool) -> Self {
+        let mut updaters: Vec<Box<dyn RepositoryForge + Send + Sync>> = Vec::new();
+        if let Ok(Some(updater)) = GitHub::new(&config) {
+            updaters.push(Box::new(updater));
+        }
+        if let Ok(updater) = GitLab::new("gitlab.com", &config.gitlab_accesstoken) {
+            updaters.push(Box::new(updater));
+        }
+        if let Ok(updater) = GitLab::new("gitlab.freedesktop.org", &None) {
+            updaters.push(Box::new(updater));
+        }
+        Self { updaters, pool }
+    }
+
+    pub(crate) fn load_repository(&self, metadata: &MetadataPackage) -> Result<Option<i32>> {
+        let url = match &metadata.repository {
+            Some(url) => url,
+            None => {
+                debug!("did not collect stats as no repository URL was present");
+                return Ok(None);
+            }
+        };
+        let mut conn = self.pool.get()?;
+        self.load_repository_inner(&mut conn, url)
+    }
+
+    fn load_repository_inner(&self, conn: &mut Client, url: &str) -> Result<Option<i32>> {
+        let name = match repository_name(url) {
+            Some(name) => name,
+            None => return Ok(None),
+        };
+
+        // Avoid querying the APIs for repositories we already loaded.
+        if let Some(row) = conn.query_opt(
+            "SELECT id FROM repositories WHERE name = $1 AND host = $2 LIMIT 1;",
+            &[&format!("{}/{}", name.owner, name.repo), &name.host],
+        )? {
+            return Ok(Some(row.get("id")));
+        }
+        for updater in self.updaters.iter().filter(|u| u.host() == name.host) {
```

Comment:
Why would there ever be more than one updater with the same host?

Reply:
No it's not. Bad relic. Replacing with a `if let` and a `filter`.